src/nodes.coffee

```diff
@@ -2402,79 +2381,108 @@ exports.Assign = class Assign extends Base
       vvar = [@makeCode ref]
       vvarText = ref
 
-    # And here comes the big loop that handles all of these cases:
-    # `[a, b] = c`
-    # `[a..., b] = c`
-    # `[..., a, b] = c`
-    # `[@a, b] = c`
-    # `[a = 1, b] = c`
-    # `{a, b} = c`
-    # `{@a, b} = c`
-    # `{a = 1, b} = c`
-    # etc.
-    for obj, i in objects
-      idx = i
-      if not expandedIdx and obj instanceof Splat
-        name = obj.name.unwrap().value
-        obj = obj.unwrap()
-        val = "#{olen} <= #{vvarText}.length ? #{utility 'slice', o}.call(#{vvarText}, #{i}"
-        rest = olen - i - 1
-        if rest isnt 0
-          ivar = o.scope.freeVariable 'i', single: true
-          val += ", #{ivar} = #{vvarText}.length - #{rest}) : (#{ivar} = #{i}, [])"
-        else
-          val += ") : []"
-        val   = new Literal val
-        expandedIdx = "#{ivar}++"
-      else if not expandedIdx and obj instanceof Expansion
-        rest = olen - i - 1
-        if rest isnt 0
-          if rest is 1
-            expandedIdx = "#{vvarText}.length - 1"
-          else
-            ivar = o.scope.freeVariable 'i', single: true
-            val = new Literal "#{ivar} = #{vvarText}.length - #{rest}"
-            expandedIdx = "#{ivar}++"
-            assigns.push val.compileToFragments o, LEVEL_LIST
-        continue
-      else
-        if obj instanceof Splat or obj instanceof Expansion
-          obj.error "multiple splats/expansions are disallowed in an assignment"
-        defaultValue = undefined
+    slicer = (type) -> (vvar, start, end = no) ->
+      args = [new IdentifierLiteral(vvar), new NumberLiteral(start)]
+      args.push new NumberLiteral end if end
+      slice = new Value new Literal "#{utility type, o}.call"
```

Comment:
Let’s avoid `new Literal`, it messes up the nodes that get output. You can achieve the same via something like `new Value (utility type, o), [new Access new PropertyName 'call']`